File: apps/ios/FitnessCoach/ViewModels/RestTimerViewModel.swift
import Foundation
import SwiftUI
import Combine

// MARK: - RestTimerViewModel
// State machine for rest timer between sets.
// States: idle → running ↔ paused → finished → idle

@MainActor
final class RestTimerViewModel: ObservableObject {

    @Published private(set) var state: RestTimerState = .idle

    private var timerTask: Task<Void, Never>?
    private let tickInterval: TimeInterval = 0.1
    var onFinished: (() -> Void)?

    // MARK: - Commands

    /// Start a new countdown from `seconds`.
    func start(seconds: TimeInterval) {
        cancel()
        state = .running(remaining: seconds, total: seconds)
        scheduleTimer()
    }

    func pause() {
        guard case .running(let r, let t) = state else { return }
        timerTask?.cancel()
        timerTask = nil
        state = .paused(remaining: r, total: t)
    }

    func resume() {
        guard case .paused(let r, let t) = state else { return }
        state = .running(remaining: r, total: t)
        scheduleTimer()
    }

    func addTime(_ delta: TimeInterval) {
        switch state {
        case .running(let r, let t):
            let newRemaining = max(1, r + delta)
            let newTotal = newRemaining > t ? newRemaining : t
            state = .running(remaining: newRemaining, total: newTotal)
        case .paused(let r, let t):
            let newRemaining = max(1, r + delta)
            let newTotal = newRemaining > t ? newRemaining : t
            state = .paused(remaining: newRemaining, total: newTotal)
        default: break
        }
    }

    func skip() {
        cancel()
        state = .idle
    }

    func cancel() {
        timerTask?.cancel()
        timerTask = nil
        state = .idle
    }

    // MARK: - Internal

    private func scheduleTimer() {
        timerTask = Task { [weak self] in
            while !Task.isCancelled {
                guard let self else { return }
                try? await Task.sleep(nanoseconds: UInt64(tickInterval * 1_000_000_000))
                if Task.isCancelled { return }

                await MainActor.run {
                    guard case .running(let remaining, let total) = self.state else {
                        self.timerTask?.cancel()
                        return
                    }
                    let next = remaining - self.tickInterval
                    if next <= 0 {
                        self.state = .finished
                        self.timerTask?.cancel()
                        self.triggerHaptic()
                        self.onFinished?()
                    } else {
                        self.state = .running(remaining: next, total: total)
                    }
                }
            }
        }
    }

    private func triggerHaptic() {
        let generator = UINotificationFeedbackGenerator()
        generator.notificationOccurred(.success)
    }
}

File: apps/android/app/src/main/java/com/fitnessapp/ui/workout/RestTimerViewModel.kt
package com.fitnessapp.ui.workout

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

// ============================================================
// RestTimerViewModel — state machine for rest timer
// States: Idle → Running ↔ Paused → Finished → Idle
// ============================================================

sealed class TimerState {
    object Idle : TimerState()
    data class Running(val remainingMs: Long, val totalMs: Long) : TimerState()
    data class Paused(val remainingMs: Long, val totalMs: Long) : TimerState()
    object Finished : TimerState()

    val isActive: Boolean get() = this is Running || this is Paused

    val remainingSeconds: Int get() = when (this) {
        is Running -> (remainingMs / 1000).toInt()
        is Paused  -> (remainingMs / 1000).toInt()
        else       -> 0
    }

    val progress: Float get() = when (this) {
        is Running -> if (totalMs > 0) 1f - remainingMs.toFloat() / totalMs.toFloat() else 0f
        is Paused  -> if (totalMs > 0) 1f - remainingMs.toFloat() / totalMs.toFloat() else 0f
        else       -> 0f
    }
}

class RestTimerViewModel : ViewModel() {

    private val _state = MutableStateFlow<TimerState>(TimerState.Idle)
    val state: StateFlow<TimerState> = _state.asStateFlow()

    private var tickJob: Job? = null
    private val tickIntervalMs = 100L

    var onFinished: (() -> Unit)? = null

    // MARK: - Commands

    fun start(seconds: Int) {
        cancel()
        val totalMs = seconds * 1000L
        _state.value = TimerState.Running(totalMs, totalMs)
        scheduleTimer()
    }

    fun pause() {
        val current = _state.value as? TimerState.Running ?: return
        tickJob?.cancel()
        tickJob = null
        _state.value = TimerState.Paused(current.remainingMs, current.totalMs)
    }

    fun resume() {
        val current = _state.value as? TimerState.Paused ?: return
        _state.value = TimerState.Running(current.remainingMs, current.totalMs)
        scheduleTimer()
    }

    fun addTime(deltaSeconds: Int) {
        val deltaMs = deltaSeconds * 1000L
        when (val current = _state.value) {
            is TimerState.Running -> {
                val newRemaining = maxOf(1000L, current.remainingMs + deltaMs)
                val newTotal = maxOf(current.totalMs, newRemaining)
                _state.value = TimerState.Running(newRemaining, newTotal)
            }
            is TimerState.Paused -> {
                val newRemaining = maxOf(1000L, current.remainingMs + deltaMs)
                val newTotal = maxOf(current.totalMs, newRemaining)
                _state.value = TimerState.Paused(newRemaining, newTotal)
            }
            else -> Unit
        }
    }

    fun skip() {
        cancel()
    }

    fun cancel() {
        tickJob?.cancel()
        tickJob = null
        _state.value = TimerState.Idle
    }

    // MARK: - Private

    private fun scheduleTimer() {
        tickJob = viewModelScope.launch {
            while (true) {
                delay(tickIntervalMs)
                val current = _state.value as? TimerState.Running ?: break
                val nextMs = current.remainingMs - tickIntervalMs
                if (nextMs <= 0) {
                    _state.value = TimerState.Finished
                    onFinished?.invoke()
                    break
                } else {
                    _state.value = TimerState.Running(nextMs, current.totalMs)
                }
            }
        }
    }

    override fun onCleared() {
        super.onCleared()
        tickJob?.cancel()
    }
}
